<a href="https://colab.research.google.com/github/jonahokiru/hello-world/blob/master/IP_Week_III_Sample.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Loading Libraries

In [ ]:
# importing pandas 
import pandas as pd

# importing numpy 
import numpy as np

# import the datetime module
import datetime as dt

## Loading Data

In [ ]:
# function to read the format of the data
def data_format():
  # number of attempts allowed to the user
  attempts = 3
  trials = 1
  while trials <= attempts:
    # user to input the data format
    data_format = input("Enter the data format:(csv or excel)\n")
    # strip user input off whitespaces and convert to lower case
    data_format = data_format.strip()
    data_format = data_format.lower()
    
    # confirm if the user to provide the correct data format
    if data_format == "csv" or data_format == "excel":
      # if the format is correct, the result of the function is the data format
      return data_format
      # End the loop 
      break
    # when the user provides wrong input
    else:
      print(f"Incorrect data format! You have {attempts - trials} left.")
    # increment the trial of the user
    trials += 1
  
  # When the user exhausts their trails 
  else:
    print("\nPlease try again later.")

In [ ]:
# Function to read the data based of the format
def load_data(path, sep, index_col):
  # use the function data_format to get the data format type
  data = data_format()
  # loading the excel data sets
  if data == "excel":
    df = pd.read_excel(path, index_col = index_col)
    return df
  # loading the csv data sets
  elif data == "csv":
    df = pd.read_csv(path, sep = sep, index_col = index_col)
    return df
  # ask the user to try again
  else:
    # user because the actions was already described in the data_format function
    pass

In [ ]:
# function with brief description the data 
def preview_data(data):

  # brief description of the data using info()
  brief_desc = data.info()

  return brief_desc

### Loading the ```cells_geo_description``` Data

In [ ]:
# pd.read_csv() has no argument sep, so passing empty space
cells_geo_description = load_data("/content/cells_geo_description.xlsx", "", None)
cells_geo_description

Enter the data format:(csv or excel)
Yves
Incorrect data format! You have 2 left.
Enter the data format:(csv or excel)
hellen
Incorrect data format! You have 1 left.
Enter the data format:(csv or excel)
     EXCel


Column name                                        Description  Format
0         VILLES                                               City  String
1         STATUS                                  In Service or not  String
2   LOCALISATION                                  in ABIDJAN or not  String
3     DECOUPZONE                                  Geographical Zone  String
4       ZONENAME                                       Name of Zone  String
5      LONGITUDE                                          Longitude   Float
6       LATITUDE                                           Latitude   Float
7         REGION                                             Region  String
8           AREA                                               Area  String
9        CELL_ID                                     ID of the cell  String
10     SITE_CODE  Site (there are several cells per site, severa...  String

* The dataset ```cells_geo_description``` contains 3 columns; **Column name**, **Description** and **Format**. 
* The dataset describes the dataset ```cells_geo```.

### Loading the ```cells_geo``` Data.

In [ ]:
# loading the cells_geo Data
cells_geo = load_data("/content/cells_geo (2).csv", ";", 0)
cells_geo.head()

Enter the data format:(csv or excel)
csf
Incorrect data format! You have 2 left.
Enter the data format:(csv or excel)
   CsV


VILLES      STATUS LOCALISATION  ... AREA     CELL_ID   SITE_CODE
0   ADJAME  In Service      ABIDJAN  ...  NaN  fbb2e75af3  0bb492ef17
1   ADJAME  In Service      ABIDJAN  ...  NaN  1502501d78  0bb492ef17
2   ADJAME  In Service      ABIDJAN  ...  NaN  0c2e182a0c  0bb492ef17
3  ASSINIE  In Service    INTERIEUR  ...  NaN  69bdfc69fd  8a1e6ce31f
4  ASSINIE  In Service    INTERIEUR  ...  NaN  7a1029d2ab  8a1e6ce31f

[5 rows x 11 columns]

In [ ]:
# Preview the data
preview_data(cells_geo)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3974 entries, 0 to 3973
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   VILLES        3974 non-null   object 
 1   STATUS        3907 non-null   object 
 2   LOCALISATION  3974 non-null   object 
 3   DECOUPZONE    3974 non-null   object 
 4   ZONENAME      3968 non-null   object 
 5   LONGITUDE     3974 non-null   float64
 6   LATITUDE      3974 non-null   float64
 7   REGION        3974 non-null   object 
 8   AREA          3951 non-null   object 
 9   CELL_ID       3974 non-null   object 
 10  SITE_CODE     3974 non-null   object 
dtypes: float64(2), object(9)
memory usage: 372.6+ KB


* The dataset ```cells_geo``` has 11 columns which have been described in the ```cells_geo_description``` dataset.
* The columns **STATUS**, **ZONENAME** and **AREA** are observed to be having missing values.

### Loading the ```CDR_description``` Data.

In [ ]:
# loading the CDR_description data
CDR_description = load_data("/content/CDR_description (1).xlsx", "", None)
CDR_description

Enter the data format:(csv or excel)
excel


Column name                                        Description   Format
0          PRODUCT                                       Voice or SMS   String
1            VALUE                                      Billing price  Integer
2        DATE_TIME               Time in format yyyy-MM-dd hh:mm:ss.0   String
3     CELL_ON_SITE  Which cell in the site was used (not needed here)  Integer
4  DW_A_NUMBER_INT  Anonymized phone number of the person for whic...   String
5  DW_B_NUMBER_INT        Anonymized phone number of the counterparty   String
6        COUNTRY_A                  Country of party A (useless here)   String
7        COUNTRY_B                  Country of party B (useless here)   String
8          CELL_ID                                     ID of the cell   String
9          SITE_ID                                     ID of the SITE   String

* The ```CDR_description``` dataset provides description for the CDR datasets.

### Loading the CDR Datasets

In [ ]:
# loading the CDR 20120507 data
CDR_20120507 = load_data("http://bit.ly/Telcom_dataset1", ",", None)
CDR_20120507.head()

Enter the data format:(csv or excel)
csv


PRODUTC  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0   Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2     sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv

[5 rows x 10 columns]

In [ ]:
# preview the dataset
preview_data(CDR_20120507)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   PRODUTC          5001 non-null   object
 1   VALUE            5001 non-null   int64 
 2   DATETIME         5001 non-null   object
 3   CELL_ON_SITE     5001 non-null   object
 4   DW_A_NUMBER_INT  5001 non-null   object
 5   DW_B_NUMBER_INT  5001 non-null   object
 6   COUNTRY_A        5001 non-null   object
 7   COUNTRY_B        5001 non-null   object
 8   CELL_ID          5001 non-null   object
 9   SITE_ID          4285 non-null   object
dtypes: int64(1), object(9)
memory usage: 390.8+ KB


In [ ]:
# loading the CDR 20120508 data
CDR_20120508 = load_data("http://bit.ly/Telcom_dataset2", ",", None)
CDR_20120508.tail()

Enter the data format:(csv or excel)
sd
Incorrect data format! You have 2 left.
Enter the data format:(csv or excel)
csv


PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
4996     sms      0  2012-05-08 00:01:53.0  ...     19e2e  896459aace  b7d00e59e4
4997     sms     25  2012-05-08 00:01:53.0  ...     19e2e  7e90b35166  211015b14d
4998     sms      0  2012-05-08 00:01:53.0  ...     19e2e  fe70e431df  c7ea989264
4999    data      0  2012-05-08 00:01:53.0  ...     OTHER  ffa6759bb2         NaN
5000   Voice      0  2012-05-08 00:01:53.0  ...     19e2e  8630058aca  b47a998a3e

[5 rows x 10 columns]

In [ ]:
# preview the dataset
preview_data(CDR_20120508)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   PRODUCT       5001 non-null   object
 1   VALUE         5001 non-null   int64 
 2   DATE_TIME     5001 non-null   object
 3   CELL_ON_SITE  5001 non-null   object
 4   DW_A_NUMBER   5001 non-null   object
 5   DW_B_NUMBER   5001 non-null   object
 6   COUNTRY_A     5001 non-null   object
 7   COUNTRY_B     5001 non-null   object
 8   CELL_ID       5001 non-null   object
 9   SITE_ID       4325 non-null   object
dtypes: int64(1), object(9)
memory usage: 390.8+ KB


In [ ]:
# loading the CDR 20120508 data
CDR_20120509 = load_data("http://bit.ly/Telcom_dataset3", ",", None)
CDR_20120509.head()

Enter the data format:(csv or excel)
csv


PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B      CELLID     SIET_ID
0   Voice     61  2012-05-08 23:01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1     sms      0  2012-05-08 23:01:31.0  ...     19e2e  +Fs4mTvdKx  +Fs4mTvdKx
2   Voice     10  2012-05-08 23:02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3     sms      0  2012-05-08 23:03:04.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     22  2012-05-08 23:03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA

[5 rows x 10 columns]

In [ ]:
# preview the dataset
preview_data(CDR_20120509)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   PRODUCT          5001 non-null   object
 1   VALUE            5001 non-null   int64 
 2   DATE_TIME        5001 non-null   object
 3   CELL_ON_SITE     5001 non-null   object
 4   DW_A_NUMBER_INT  5001 non-null   object
 5   DW_B_NUMBER_INT  5001 non-null   object
 6   COUNTRY_A        5001 non-null   object
 7   COUNTRY_B        5001 non-null   object
 8   CELLID           5001 non-null   object
 9   SIET_ID          4396 non-null   object
dtypes: int64(1), object(9)
memory usage: 390.8+ KB


In all the three datasets:
* Each dataset has **10 columns** and **5001 rows**.
* The column **SIET_ID** has missing values.
* The column **CELL_ON_SITE** is of data type ```int``` and not ```object``` as per the data description while the column **DATE_TIME** should be converted to ```datetime```.

## Data Preparation
* Rename the columns of the datasets
* Grouping the three CDR datasets into one dataset and merge with the cells_geo data.
* Checking the dates provided if they are for three days.


### Renaming the columns

In [ ]:
# changing the font 
cells_geo.columns = map(str.lower, cells_geo.columns)
# ensure that the columns names are correctly labelled according to description
cells_geo.columns

Index(['villes', 'status', 'localisation', 'decoupzone', 'zonename',
       'longitude', 'latitude', 'region', 'area', 'cell_id', 'site_code'],
      dtype='object')

In [ ]:
# ensure columns correctly labelled in the three datasets
columns = ['PRODUCT', 'VALUE', 'DATE_TIME', 'CELL_ON_SITE', 'DW_A_NUMBER_INT', 'DW_B_NUMBER_INT', 'COUNTRY_A', 'COUNTRY_B', 'CELL_ID', 'SITE_ID']
data_sets = [CDR_20120507, CDR_20120508, CDR_20120509]

# use a loop to assign the three datasets similar column names and change the case
for data in data_sets:
  # assign similar column name s
  data.columns = columns
  # change case
  data.columns = map(str.lower, data.columns)

### Grouping the data 
* Concat the CDR datasets into one large dataset
* Merge the CDR dataset with the cells_geo dataset

In [ ]:
# concat the three CDR data frames
CDR = pd.concat([CDR_20120507,CDR_20120508,CDR_20120509])
# resetting the indexes, drop the old indexes
CDR.reset_index(drop=True, inplace=True)
# check the shape
CDR.shape

(15003, 10)

In [ ]:
# display the names of the cells geo dataset
cells_geo.columns

Index(['villes', 'status', 'localisation', 'decoupzone', 'zonename',
       'longitude', 'latitude', 'region', 'area', 'cell_id', 'site_code'],
      dtype='object')

In [ ]:
# display the columns names of the CDR data
CDR.columns


Index(['product', 'value', 'date_time', 'cell_on_site', 'dw_a_number_int',
       'dw_b_number_int', 'country_a', 'country_b', 'cell_id', 'site_id'],
      dtype='object')

In [ ]:
# function to check if values in one column exit as values in another column
def similar_columns(data1,data2,col1,col2):
  # strip of the white spaces if any
  data1[col1] = data1[col1].str.strip()
  data2[col2] = data2[col2].str.strip()
  # convert columns in both datasets into list.
  first_list = data1[col1].to_list()
  second_list = data2[col2].to_list()

  # check if any values appears in both lists
  check = any(item in first_list for item in second_list)

  return check

In [ ]:
# check if the values of cell_id are on both CDR and cells_geo data set
similar_columns(CDR, cells_geo, 'cell_id', 'cell_id')

False

In [ ]:
# check if the values of site_code of cells geo are in column site_id in CDR
similar_columns(CDR, cells_geo, 'site_id', 'site_code')

True

In [ ]:
# merge the two datasets
df = pd.merge(cells_geo, CDR, how ='inner', left_on='site_code', right_on='site_id')
df.head()

villes      status localisation  ... country_b   cell_id_y     site_id
0  ADJAME  In Service      ABIDJAN  ...     19e2e  ac8f1ccb39  0bb492ef17
1  ADJAME  In Service      ABIDJAN  ...     19e2e  ac8f1ccb39  0bb492ef17
2  ADJAME  In Service      ABIDJAN  ...     19e2e  ac8f1ccb39  0bb492ef17
3   ABOBO  In Service      ABIDJAN  ...     19e2e  35a513722e  56306cc7ca
4   ABOBO  In Service      ABIDJAN  ...     19e2e  35a513722e  56306cc7ca

[5 rows x 21 columns]

In [ ]:
# check the preview of the dataset
preview_data(df)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 51361 entries, 0 to 51360
Data columns (total 21 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   villes           51361 non-null  object 
 1   status           50645 non-null  object 
 2   localisation     51361 non-null  object 
 3   decoupzone       51361 non-null  object 
 4   zonename         51343 non-null  object 
 5   longitude        51361 non-null  float64
 6   latitude         51361 non-null  float64
 7   region           51361 non-null  object 
 8   area             51105 non-null  object 
 9   cell_id_x        51361 non-null  object 
 10  site_code        51361 non-null  object 
 11  product          51361 non-null  object 
 12  value            51361 non-null  int64  
 13  date_time        51361 non-null  object 
 14  cell_on_site     51361 non-null  object 
 15  dw_a_number_int  51361 non-null  object 
 16  dw_b_number_int  51361 non-null  object 
 17  country_a   

**Conclusions:**
* The CDR datasets were concatenated to form one dataset as they were similar, but distributed across different time zones.
* The CDR and cells_geo datasets could not be merged on the ```cells_id``` column as there was no any entry in the cells_geo data on the CDR data.
* There are similar values on the column ```site_code``` on ```cells_geo``` and the column ```site_id``` on the ```CDR``` datasets which can be used to merge the dataset into one.

### Confirming dates

In [ ]:
# function to confirm dates
def confirm_dates(data, col):
  # convert the datetime column from string to datetime
  data[col] = data[col].apply(lambda x: 
                              dt.datetime.strptime(x, '%Y-%m-%d %H:%M:%S.%f'))
  # get the day from the timestamp 
  data['DAY'] = pd.to_datetime(data[col]).dt.day
  # get the unique value counts in the day column to confirm the date
  output = data['DAY'].value_counts()
  # drop the column
  data.drop('DAY', axis = 1, inplace = True)

  # get the unique day(s)
  return output

In [ ]:
# confirming the days on the joined dataset
confirm_dates(df, "date_time")

9    18457
7    16610
8    16294
Name: DAY, dtype: int64

In [ ]:
# getting the maximum and minimum time 
def max_min_time(data, col):
  min_time = min(data[col])
  max_time = max(data[col])
  
  return min_time, max_time 

In [ ]:
# confirming the durations
max_min_time(df, "date_time")

(Timestamp('2012-05-07 00:00:00'), Timestamp('2012-05-09 00:01:43'))

**Conclusions:**
* After joining the four datasets, the data had dates **7th**, **8th** and **9th**.






## Data Cleaning
* Dropping unnecessary columns.
* Checking and handling missing values.
* Removing duplicates if any.
* Handling outliers.

### Dropping Unnecessary Columns

In [ ]:
# preview data columns
df.head()

villes      status localisation  ... country_b   cell_id_y     site_id
0  ADJAME  In Service      ABIDJAN  ...     19e2e  ac8f1ccb39  0bb492ef17
1  ADJAME  In Service      ABIDJAN  ...     19e2e  ac8f1ccb39  0bb492ef17
2  ADJAME  In Service      ABIDJAN  ...     19e2e  ac8f1ccb39  0bb492ef17
3   ABOBO  In Service      ABIDJAN  ...     19e2e  35a513722e  56306cc7ca
4   ABOBO  In Service      ABIDJAN  ...     19e2e  35a513722e  56306cc7ca

[5 rows x 21 columns]

In [ ]:
# getting the correct column names
df.columns

Index(['villes', 'status', 'localisation', 'decoupzone', 'zonename',
       'longitude', 'latitude', 'region', 'area', 'cell_id_x', 'site_code',
       'product', 'value', 'date_time', 'cell_on_site', 'dw_a_number_int',
       'dw_b_number_int', 'country_a', 'country_b', 'cell_id_y', 'site_id'],
      dtype='object')

In [ ]:
# creating a function to drop unnecessary columns
def drop_columns(data, cols):
  # drop the columns not required
  data.drop(cols, axis = 1, inplace = True)

  return data


# columns to drop
columns = ['longitude', 'latitude', 'cell_id_x', 'cell_on_site', 'country_a', 'country_b', 'cell_id_y']
df = drop_columns(df, columns)

**Conclusion:**
* The columns ```longituted``` & ```latitude``` are used to provide location which is already provided in the columns ```villes``` and ```localisation```.
* The columns ```cell_id_x``` and ```cell_id_y``` are the ID of the cell which is not required in the study.
* The columns ```country_a```, ```country_b``` and ```cell_on_site``` are described as not necessary in the documentation.

### Checking and Handling missing values


In [ ]:
# creating a function to check 
def missing_values(data):

  # identify the missing values in the train_df Dataframe
  #
  miss = data.isnull().sum().sort_values(ascending = False)
  # the percentage of missing value 
  #
  percentage = (data.isnull().sum() / len(data)).sort_values(ascending = False)
  # storing total missing value aganist percentage in a Dataframe
  #
  missing_data = pd.DataFrame({'Total': miss, 'Percentage': percentage})
  # if the percentage is 0, indicates no missing values hence removed
  # 
  missing_data.drop(missing_data[missing_data['Percentage'] == 0].index, inplace = True)
  # diplaying the dataframe with the missing value
  #
  return missing_data



missing_values(df)

Total  Percentage
status      716    0.013941
area        256    0.004984
zonename     18    0.000350

* The columns ```status```, ```area```, and ```zonename``` all have missing values.
* This accounts for less than **1%** of the data, hence can be dropped.

In [ ]:
# function to drop missing values
def drop_na(df):
  